crates/turborepo-lib/src/task_hash.rs

```diff
@@ -0,0 +1,318 @@
+        global_hash: &'a str,
+    ) -> Self {
+        Self {
+            opts,
+            env_at_execution_start,
+            global_hash,
+            task_hash_tracker: Mutex::new(TaskHashTracker::default()),
+        }
+    }
+
+    /// Calculates the package inputs hash, i.e. the hash of
+    /// the files for a given task.
+    /// This is a little bit of an expensive operation,
+    /// so in the future we should look into caching it.
+    pub async fn calculate_package_inputs_hash(
+        &self,
+        scm: &SCM,
+        task_id: &TaskId<'static>,
+        task_definition: &TaskDefinition,
+        pkg: &WorkspaceInfo,
+        repo_root: &AbsoluteSystemPath,
+    ) -> Result<Option<String>, Error> {
+        if task_id.package() == ROOT_PKG_NAME {
+            return Ok(None);
+        }
+
+        let workspace_name = WorkspaceName::Other(task_id.package().to_string());
+
+        let package_file_hash_inputs = PackageFileHashInputs {
+            task_id: task_id.clone(),
+            task_definition,
+            workspace_name,
+        };
+
+        let package_path = pkg
+            .package_json_path
+            .parent()
+            .unwrap_or_else(|| AnchoredSystemPath::new("").unwrap());
+
+        let mut hash_object = scm.get_package_file_hashes(
+            &repo_root,
+            package_path,
+            &package_file_hash_inputs.task_definition.inputs,
+        )?;
+
+        if !package_file_hash_inputs.task_definition.dot_env.is_empty() {
+            let package_path = pkg
+                .package_json_path
+                .parent()
+                .unwrap_or_else(|| AnchoredSystemPath::new("").unwrap());
+            let absolute_package_path = repo_root.resolve(package_path);
+            let dot_env_object = scm.hash_existing_of(
+                &absolute_package_path,
+                package_file_hash_inputs
+                    .task_definition
+                    .dot_env
+                    .iter()
+                    .map(|p| p.to_anchored_system_path_buf()),
+            )?;
+
+            for (key, value) in dot_env_object {
+                hash_object.insert(key, value);
+            }
+        }
+
+        let file_hashes = FileHashes(hash_object);
+        let hash = file_hashes.clone().hash();
+
+        Ok(Some(hash))
+    }
+
+    pub async fn calculate_task_hash(
+        &self,
+        scm: &SCM,
+        task_id: &TaskId<'static>,
+        task_definition: &TaskDefinition,
+        task_env_mode: ResolvedEnvMode,
+        workspace: &WorkspaceInfo,
+        dependency_set: HashSet<&TaskNode>,
+        repo_root: &AbsoluteSystemPath,
+    ) -> Result<String, Error> {
+        let do_framework_inference = self.opts.run_opts.framework_inference;
+        let is_monorepo = !self.opts.run_opts.single_package;
+
+        let hash_of_files = self
+            .calculate_package_inputs_hash(scm, task_id, task_definition, workspace, repo_root)
```

Comment:
Is there a reason for the change from the Go behavior where we eagerly calculate this?

I think it would be beneficial to keep that behavior as it allows us to easily use something like rayon to parallelize the calculation of these inputs.

Reply:
I'm not sure rayon is the right tool here since the hashes are computed by shelling out to git which is probably not CPU bound. I didn't want to spin up multiple concurrency primitives here because it felt unnecessary when we could use the visitor for both